src/statsmodels/formula.jl

```diff
@@ -444,59 +455,79 @@ creating the model matrix.
+                          view(terms.eterms, f),
+                          view(terms.is_non_redundant, f, i_term))
+        num_term_cols = prod(block_sizes)
+
+        push!(term_col_inds, (1:num_term_cols)+num_cols)
+        num_cols += num_term_cols
+    end
+
+    if num_cols == 0
+        error("Could not construct model matrix. Resulting matrix has 0 columns.")
+    end
+
+    # Pre-allocate model matrix
+    mm = similar(convert(T, Matrix()), size(dfrm, 1), num_cols)
+    assign = zeros(Int, num_cols)
 
-    blocks = T[]
-    assign = Int[]
     if terms.intercept
-        push!(blocks, ones(size(dfrm, 1), 1))  # columns of 1's is first block
-        push!(assign, 0)                       # this block corresponds to term zero
+        mm[:,1] = 1
     end
 
-    factors = terms.factors
+    ## Step 2:
+    ## Fill in the columns for each term.
 
     ## Map eval. term name + redundancy bool to cached model matrix columns
+    ## (views of model matrix itself for eval terms that correspond directly to
+    ## mm columns, and arrays otherwise)
     eterm_cols = @compat Dict{Tuple{Symbol,Bool}, T}()
-    ## Accumulator for each term's vector of eval. term columns.
-
-    ## TODO: this method makes multiple copies of the data in the ModelFrame:
-    ## first in term_cols (1-2x per evaluation term, depending on redundancy),
-    ## second in constructing the matrix itself.
 
     ## turn each term into a vector of mm columns for its eval. terms, using
     ## "promoted" full-rank versions of categorical columns for non-redundant
     ## eval. terms:
     for (i_term, term) in enumerate(terms.terms)
-        term_cols = T[]
         ## Pull out the eval terms, and the non-redundancy flags for this term
-        ff = Compat.view(factors, :, i_term)
-        eterms = Compat.view(terms.eterms, ff)
-        non_redundants = Compat.view(terms.is_non_redundant, ff, i_term)
-        ## Get cols for each eval term (either previously generated, or generating
-        ## and storing as necessary)
-        for (et, nr) in zip(eterms, non_redundants)
-            if ! haskey(eterm_cols, (et, nr))
-                eterm_cols[(et, nr)] = modelmat_cols(T, et, mf, non_redundant=nr)
+        ff = view(factors, :, i_term)
+        eterms = view(terms.eterms, ff)
+        non_redundants = view(terms.is_non_redundant, ff, i_term)
+        if length(eterms) == 1 && term == eterms[1]
+            ## handle special case where term == eterm (e.g., main effect
+            ## terms). in this case, we can store the columns generated for the
+            ## eterm directly in the model matrix and just hold onto a view of
+            ## those columns for any higher-order terms that need them
+            et = eterms[1]
+            nr = non_redundants[1]
+            mm[:, term_col_inds[i_term]] = modelmat_cols(T, et, mf, non_redundant=nr)
+            eterm_cols[(et, nr)] = view(mm, :, term_col_inds[i_term])
```

Comment:
It looks like the extra allocations are happening here: by one benchmark with 24Mb of allocations for the actual model matrix, there are 16Mb of allocations on this line.